File: src/main/java/com/puhovin/intellijplugin/twm/settingsmanager/AbstractSettingsManager.java
package com.puhovin.intellijplugin.twm.settingsmanager;

import com.puhovin.intellijplugin.twm.model.ToolWindowPreference;
import com.puhovin.intellijplugin.twm.model.ToolWindowPreferenceStore;
import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.Map;

public abstract class AbstractSettingsManager implements SettingsManager {

    protected ToolWindowPreferenceStore state = new ToolWindowPreferenceStore();
    protected final Map<String, ToolWindowPreference> defaultPreferences = new HashMap<>();

    @Override
    public @NotNull ToolWindowPreferenceStore getState() {
        return state;
    }

    @Override
    public void loadState(@NotNull ToolWindowPreferenceStore state) {
        this.state = state;
        applyPreferences(state.getPreferences());
    }

    @Override
    public Map<String, ToolWindowPreference> getDefaultPreferences() {
        return defaultPreferences;
    }

    @Override
    public void setDefaultPreferences(@NotNull Map<String, ToolWindowPreference> defaultPreferences) {
        this.defaultPreferences.putAll(defaultPreferences);
    }

    @Override
    public void applyPreferences(@NotNull Map<String, ToolWindowPreference> preferences) {
        this.state.setPreferences(preferences);
    }

    @Override
    public ToolWindowPreference getDefaultAvailabilityToolWindow(String id) {
        return defaultPreferences.get(id);
    }
}

File: src/main/kotlin/com/puhovin/intellijplugin/twm/settingsmanager/AbstractSettingsManager.kt
package com.puhovin.intellijplugin.twm.settingsmanager

import com.puhovin.intellijplugin.twm.model.ToolWindowPreference
import com.puhovin.intellijplugin.twm.model.ToolWindowPreferenceStore

abstract class AbstractSettingsManager : SettingsManager {

    private var state = ToolWindowPreferenceStore()
    private val defaultPreferences = mutableMapOf<String, ToolWindowPreference>()

    override fun getState(): ToolWindowPreferenceStore {
        return state
    }

    override fun loadState(state: ToolWindowPreferenceStore) {
        this.state = state
    }

    override fun getPreferences(): Map<String, ToolWindowPreference> {
        return state.getPreferences()
    }

    override fun setPreferences(preferences: Map<String, ToolWindowPreference?>) {
        state.setPreferences(preferences)
    }

    override fun getDefaultPreferences(): Map<String, ToolWindowPreference> {
        return defaultPreferences
    }

    override fun setDefaultPreferences(preferences: Map<String, ToolWindowPreference>) {
        this.defaultPreferences.putAll(preferences)
    }

    override fun getDefaultAvailabilityToolWindow(id: String?): ToolWindowPreference? {
        return defaultPreferences[id]
    }
}